src/codec/lz4/encoder.rs

```diff
@@ -0,0 +1,301 @@
+}
+
+impl EncoderContext {
+    fn new() -> Result<Self> {
+        let mut context = LZ4FCompressionContext(core::ptr::null_mut());
+        check_error(unsafe { LZ4F_createCompressionContext(&mut context, LZ4F_VERSION) })?;
+        Ok(Self { ctx: context })
+    }
+}
+
+impl Drop for EncoderContext {
+    fn drop(&mut self) {
+        unsafe { LZ4F_freeCompressionContext(self.ctx) };
+    }
+}
+
+impl Lz4Encoder {
+    pub(crate) fn new(preferences: LZ4FPreferences) -> Self {
+        let block_size = preferences.frame_info.block_size_id.get_size();
+
+        let block_buffer_size = min_dst_size(block_size, &preferences);
+        let flush_buffer_size = min_dst_size(0, &preferences);
+
+        Self {
+            ctx: Unshared::new(EncoderContext::new().unwrap()),
+            state: State::Header,
+            preferences,
+            limit: block_size,
+            maybe_buffer: None,
+            block_buffer_size,
+            flush_buffer_size,
+        }
+    }
+
+    pub(crate) fn buffer_size(&self) -> usize {
+        self.block_buffer_size
+    }
+
+    fn drain_buffer(
+        &mut self,
+        output: &mut PartialBuffer<impl AsRef<[u8]> + AsMut<[u8]>>,
+    ) -> (usize, usize) {
+        match self.maybe_buffer.as_mut() {
+            Some(buffer) => {
+                let drained_bytes = output.copy_unwritten_from(buffer);
+                (drained_bytes, buffer.unwritten().len())
+            }
+            None => (0, 0),
+        }
+    }
+
+    fn write<'a, T>(
+        &'a mut self,
+        lz4_fn: Lz4Fn<'a, T>,
+        output: &'a mut PartialBuffer<impl AsRef<[u8]> + AsMut<[u8]>>,
+    ) -> Result<usize>
+    where
+        T: AsRef<[u8]>,
+    {
+        let (drained_before, undrained) = self.drain_buffer(output);
+        if undrained > 0 {
+            return Ok(drained_before);
+        }
+
+        let min_dst_size = match &lz4_fn {
+            Lz4Fn::Begin => LZ4F_HEADER_SIZE_MAX,
+            Lz4Fn::Update { input } => {
+                min_dst_size(input.unwritten().len().min(self.limit), &self.preferences)
+            }
+            Lz4Fn::Flush | Lz4Fn::End => self.flush_buffer_size,
+        };
+
+        let (dst_buffer, maybe_internal_buffer) = if min_dst_size > output.unwritten().len() {
+            let buffer_size = self.block_buffer_size;
+            let buffer = self
+                .maybe_buffer
+                .get_or_insert_with(|| PartialBuffer::new(Vec::with_capacity(buffer_size)));
+            buffer.reset();
+            (buffer.unwritten_mut().as_mut_ptr(), Some(buffer))
+        } else {
+            (output.unwritten_mut().as_mut_ptr(), None)
+        };
+
+        let len = match lz4_fn {
+            Lz4Fn::Begin => {
+                let len = check_error(unsafe {
+                    LZ4F_compressBegin(
+                        self.ctx.get_mut().ctx,
+                        dst_buffer,
+                        min_dst_size,
```

Comment:
```suggestion
                        dst_buffer.unwritten().len(),
```

Reply:
`dst_buffer` is a mut pointer now, so this is not possible. `min_dst_size` is the worst case size of the destination buffers required by the [lz4 C functions](https://github.com/lz4/lz4/blob/9d53d8bb6c4120345a0966e5d8b16d7def1f32c5/lib/lz4frame.h#L306-L320), hence it is safe to use here instead of the actual destination buffer size.

Reply:
We could evaluate its length before the function call, then there would only be one active mutable borrow